experiment/simulation/js/main.js

```diff
@@ -1 +1,579 @@
-//Your JavaScript goes in here
+const submit_button = document.getElementById("submit");
+const next_button = document.getElementById("next-button");
+const prev_button = document.getElementById("prev-button");
+const reset_button = document.getElementById("reset-button");
+const formx1 = document.getElementById("x1");
+const formy1 = document.getElementById("y1");
+const formx2 = document.getElementById("x2");
+const formy2 = document.getElementById("y2");
+const errorx1 = document.getElementById("error-x1");
+const errorx2 = document.getElementById("error-x2");
+const errory1 = document.getElementById("error-y1");
+const errory2 = document.getElementById("error-y2");
+// get the canvas
+
+let canvas = document.getElementById("canvas");
+// get the context
+let ctx = canvas.getContext("2d");
+const height = 600;
+const width = 1200;
+canvas.height = height;
+canvas.width = width;
+
+const chosen_color = "green";
+const possible_color = "blue";
+// determine the number of divisions based on the distance between the two points
+const divisions = 45;
+const MAXX = 88;
+const MINX = -100;
+const MAXY = 44;
+const MINY = 0;
+const grid_color = "white";
+const originx = 45;
+const originy = 23;
+const block_size = height / divisions;
+
+// define the set of coordinates depict them as 2-d vectors
+let point1 = [];
+let point2 = [];
+let decision_parameter;
+let dp = [];
+// fix the heights of the canvas
+let slope = 0;
+let currx, curry;
+let dx, dy;
+let last_move_direction = "";
+let valid = 1;
+
+// let submit_button = document.getElementById("submit");
+// store the number of times next is called
+let times_next_called = 0;
+
+formx1.onchange = function () {
+  let x1 = formx1.value;
+  if (x1 < -45 || x1 > 44) {
+    valid = 0;
+    formx1.style.border = "1px solid red";
+    errorx1.innerHTML = "x should be between " + "-45" + " and " + "44";
+  } else {
+    valid = 1;
+    formx1.style.border = "none";
+    errorx1.innerHTML = "";
+  }
+};
+formy1.onchange = function () {
+  let y1 = formy1.value;
+  if (y1 < -23 || y1 > 21) {
+    valid = 0;
+    formy1.style.border = "1px solid red";
+    errory1.innerHTML = "y should be between " + "-23" + " and " + "21";
+  } else {
+    valid = 1;
+    formy1.style.border = "none";
+
+    errory1.innerHTML = "";
+  }
+};
+formx2.onchange = function () {
+  let x2 = formx2.value;
+  if (x2 < -45 || x2 > 44) {
+    formx2.style.border = "1px solid red";
+    valid = 0;
+    errorx2.innerHTML = "x should be between " + "-45" + " and " + "44";
+  } else {
+    valid = 1;
+    formx2.style.border = "none";
+   
+    errorx2.innerHTML = "";
+  }
+};
+formy2.onchange = function () {
```

Comment:
Do you really need 4 functions to validate formx and formy?
Can you think of creating one function to validate input, pass in the element to be validated and a list of validation conditions? That way it'd be a generic function and you can apply all styling in one place after considering multiple validations on the element.
If that sounds too generic and a bit complicated, please write one function and use it to validate all of formx and formy

Reply:
one way it could be done is , when the user clicks on submit , then it validates for all the form fields at once , that could be implemented , but as per the current implementation it displays the errors alongside the user fills one of the form fields , i.e. without clicking on submit button. If that works fine , else we could implement it under one hood on clicking submit button maybe ?